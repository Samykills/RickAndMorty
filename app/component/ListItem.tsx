import React, { PureComponent } from 'react';
import { View, Dimensions, Animated, Text, Image, TouchableWithoutFeedback } from 'react-native';
import PropTypes from 'prop-types';

const ITEM_WIDTH = Dimensions.get('window').width;
const ANIMATION_DURATION = 250;

class ListItem extends PureComponent<Props> {
  listItemAnimation = new Animated.Value(0);

  onPressInEffect = () => {
    Animated.timing(this.listItemAnimation, {
      toValue: 1,
      duration: ANIMATION_DURATION,
      useNativeDriver: true
    }).start();
  };

  onPressOutEffect = () => {
    Animated.timing(this.listItemAnimation, {
      toValue: 0,
      duration: ANIMATION_DURATION,
      useNativeDriver: true
    }).start();
  };

  render() {
    const { item, numColumns } = this.props;
    const listItemScale = this.listItemAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [1, 1.2],
      extrapolate: 'clamp'
    });

    const listItemImageOpacity = this.listItemAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [0.8, 1],
      extrapolate: 'clamp'
    });

    const listItemTextOpacity = this.listItemAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [1, 0],
      extrapolate: 'clamp'
    });

    const CALC_ITEM_WIDTH = (ITEM_WIDTH - 20 * numColumns) / numColumns;

    return (
      <TouchableWithoutFeedback onPressIn={this.onPressInEffect} onPressOut={this.onPressOutEffect}>
        <Animated.View
          style={{
            margin: 10,
            transform: [
              {
                scale: listItemScale
              }
            ]
          }}
        >
          <Animated.Image
            style={{
              height: 100,
              width: CALC_ITEM_WIDTH,
              borderRadius: 5,
              opacity: listItemImageOpacity
            }}
            resizeMode='cover'
            source={{
              uri: item.image
            }}
          />
          <Animated.View
            style={{
              position: 'absolute',
              top: '50%',
              backgroundColor: '#FAFAFAA0',
              opacity: listItemTextOpacity,
              width: CALC_ITEM_WIDTH
            }}
          >
            <Text
              style={{
                textAlign: 'center',
                fontWeight: 'bold'
              }}
              ellipsizeMode='tail'
              numberOfLines={2}
            >
              {item.name}
            </Text>
            <Text
              style={{
                textAlign: 'center',
                fontSize: 9,
                color: item.status == 'Alive' ? 'green' : 'red'
              }}
            >
              {item.status}
            </Text>
          </Animated.View>
        </Animated.View>
      </TouchableWithoutFeedback>
    );
  }
}

export interface Props {
  item: object;
  numColumns: number;
}
export default ListItem;
