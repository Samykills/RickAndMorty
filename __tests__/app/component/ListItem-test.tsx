// __tests__/Intro-test.js
import React from 'react';
import ListItem from '../../../app/component/ListItem';
import renderer from 'react-test-renderer';

const ITEM = { image: 'https://comicsalliance.com/files/2016/01/rick-and-morty.jpg', name: 'Rick', status: 'Alive' };
const NUM_COLUMNS = 3;
test('ListItem renders correctly', () => {
  const tree = renderer.create(<ListItem item={ITEM} numColumns={NUM_COLUMNS} />).toJSON();
  expect(tree).toMatchSnapshot();
});
