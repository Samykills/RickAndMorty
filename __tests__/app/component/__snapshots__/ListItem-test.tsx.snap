// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "margin": 10,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "https://comicsalliance.com/files/2016/01/rick-and-morty.jpg",
      }
    }
    style={
      Object {
        "borderRadius": 5,
        "height": 100,
        "opacity": 0.8,
        "width": 230,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#FAFAFAA0",
        "opacity": 1,
        "position": "absolute",
        "top": "50%",
        "width": 230,
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={2}
      style={
        Object {
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Rick
    </Text>
    <Text
      style={
        Object {
          "color": "green",
          "fontSize": 9,
          "textAlign": "center",
        }
      }
    >
      Alive
    </Text>
  </View>
</View>
`;
